deepmd/env.py

```diff
@@ -107,22 +107,30 @@ def get_tf_default_nthreads() -> Tuple[int, int]:
     )
 
 
-def get_tf_session_config() -> Any:
+def get_tf_session_config(cpu_only: bool = False) -> Any:
     """Configure tensorflow session.
 
+    Parameters
+    ----------
+    cpu_only : bool, optional
+        If enabled, no GPU device is visible to the TensorFlow Session.
+
     Returns
     -------
     Any
         session configure object
     """
     set_tf_default_nthreads()
     intra, inter = get_tf_default_nthreads()
-    return tf.ConfigProto(
+    config = tf.ConfigProto(
         intra_op_parallelism_threads=intra, inter_op_parallelism_threads=inter
     )
+    if cpu_only:
+        config.device_count['GPU'] = 0
+    return config
 
 
-default_tf_session_config = get_tf_session_config()
+default_tf_session_config = get_tf_session_config(cpu_only=True)
```

Comment:
Currently, there is no way to know which command (train/test/compress...) is activated in the scope of `descriptor`, `neighbor_stat`, `type_embedding` and `infer.*`.
I suggested defining a  global object to tell context. As a result, each class holding a TensorFlow session can switch to correct CPU or GPU config without conflict among multiple sessions.
Talked with @denghuilu , he recommends to disable GPU usage for non-training case temporarily and we can design a better mechanism to manage session (maybe singleton mode) in another pull request.

Reply:
The confliction error of multiple sessions in one process can be reproduced using the script below. @amcadmus 
```python
# Suppose there are 2 GPU cards visible to current process
import tensorflow as tf

config = tf.ConfigProto()  # default_tf_session_config
sub_sess = tf.Session()

config.gpu_options.visible_device_list = u'0'
training_sess = tf.Session(config)
```

Here comes the error message:
```text
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.6/site-packages/tensorflow_core/python/client/session.py", line 1585, in __init__
    super(Session, self).__init__(target, graph, config=config)
  File "/usr/local/lib/python3.6/site-packages/tensorflow_core/python/client/session.py", line 699, in __init__
    self._session = tf_session.TF_NewSessionRef(self._graph._c_graph, opts)
tensorflow.python.framework.errors_impl.AlreadyExistsError: TensorFlow device (GPU:0) is being mapped to multiple CUDA devices (1 now, and 0 previously), which is not supported. This may be the result of providing different GPU configurations (ConfigProto.gpu_options, for example different visible_device_list) when creating multiple Sessions in the same process. This is not  currently supported, see https://github.com/tensorflow/tensorflow/issues/19083
```

Reply:
The confliction of multiple sessions in one process can be reproduced using the script below. @amcadmus 
```python
# Suppose there are 2 GPU cards visible to current process
import tensorflow as tf

config = tf.ConfigProto()  # default_tf_session_config
sub_sess = tf.Session()

config.gpu_options.visible_device_list = u'1'
training_sess = tf.Session(config=config)
```

Here comes the error message:
```text
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.6/site-packages/tensorflow_core/python/client/session.py", line 1585, in __init__
    super(Session, self).__init__(target, graph, config=config)
  File "/usr/local/lib/python3.6/site-packages/tensorflow_core/python/client/session.py", line 699, in __init__
    self._session = tf_session.TF_NewSessionRef(self._graph._c_graph, opts)
tensorflow.python.framework.errors_impl.AlreadyExistsError: TensorFlow device (GPU:0) is being mapped to multiple CUDA devices (1 now, and 0 previously), which is not supported. This may be the result of providing different GPU configurations (ConfigProto.gpu_options, for example different visible_device_list) when creating multiple Sessions in the same process. This is not  currently supported, see https://github.com/tensorflow/tensorflow/issues/19083
```

Reply:
The command `dp test` calls the python inference interface like `deepmd.infer.DeepPot`, which are computationally demanding. These interfaces may also be called by users in their python code. Thus detection of GPUs for the inference is important. What we assume is that the inference is serial, i.e. if there exists GPUs, it only use the first of them. 

My point is that the implementation of parallel training should not change the default behavior of using devices in other parts of the code.

Reply:
@amcadmus How about adding a `mode` property in `deepmd.env`？When `dp train` is executed, we set `mode=TRAIN`. In consequence, all sub sessions use CPU only.

Reply:
Can we use an environmental variable to control the behavior?

Reply:
Does HOROVOD have existing environment variables? If so, we don't need to define extra variables.

Reply:
@njzjz Your suggesion should be OK as well.
Finally, I choose the way that rewrite the default session config at the begining of distributed training.

Reply:
> Can we use an environmental variable to control the behavior?

In this PR we keep it as is, and consider this option in a separate PR